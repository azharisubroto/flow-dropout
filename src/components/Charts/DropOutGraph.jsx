import React from 'react'
import VerticalBarChart from './VerticalBarChart'
import DiffIndicator from './DiffIndicator'
import PropTypes from 'prop-types'
import styles from './DropOutGraph.module.scss'

/**
 * Render DropOutGraph
 *
 * @author Azhari Subroto <azhari.subroto@gmail.com>
 * @param {Object} props
 * @component
 * @example
 * return (
 *   <DropOutGraph jsonData={json_object} height={450} />
 * )
 */
const DropOutGraph = ({ jsonData, height, className, ...other }) => {
  const items = jsonData
  const keys = Object.keys(items).sort()

  /**
   * Get Value diff
   * @param {String} item - Item name from Object
   * @param {Object} currentVal - Value of an current item
   */
  const getDiff = (item, currentVal) => {
    const prevVal = currentVal
    const loc = parseInt(keys.indexOf(item) + 1)
    const nextitem = items[keys[loc]]
    const nextVal = nextitem ? nextitem.value : 0
    const diff = parseFloat(Math.abs(prevVal - nextVal).toFixed(2))
    const diffWithSymbol = `${prevVal > nextVal ? '-' : '+'}${diff}`

    if (nextVal == 0) {
      return null
    }

    return `${diff != 0 ? diffWithSymbol : 0}`
  }

  // classes
  let classes = styles.chartcontainer
  if (className) {
    classes += ` ${className}`
  }

  return (
    <div className={classes} {...other} style={{ height: `${height ? height : 500}px` }}>
      {/* Object Looping */}
      {Object.entries(items).map(([item, val], i) => {
        const diff = getDiff(item, val.value)

        return (
          <React.Fragment key={`bar-${i}`}>
            <VerticalBarChart
              label={val.label}
              item={item}
              barValue={val.value}
              type={val.type}
              adjList={val.adjList}
            />

            {/* Value Diff */}
            {diff != null && <DiffIndicator diffValue={parseInt(diff)} />}
          </React.Fragment>
        )
      })}
    </div>
  )
}

// Validate prop types
DropOutGraph.propTypes = {
  /** - Json Data */
  jsonData: PropTypes.object,
  /** - Chart Height */
  height: PropTypes.number
}

DropOutGraph.defaultProps = {
  height: 450
}

export default DropOutGraph
