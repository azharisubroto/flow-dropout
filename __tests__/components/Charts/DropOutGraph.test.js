import React from 'react'
import { shallow } from 'enzyme'
import DropOutGraph from '@/components/Charts/DropOutGraph'

describe('<DropOutGraph />', () => {
  let wrapper, DropOutGraphComp

  beforeEach(() => {
    const NodeMap = {
      node1: {
        label: 'Contact Info',
        value: 100,
        type: 'BASIC',
        adjList: ['node2']
      },
      node2: {
        label: 'NemID',
        value: 95,
        type: 'SERVICE',
        adjList: ['node3', 'nodeBranch2']
      },
      node3: {
        label: 'Personal Address',
        value: 95,
        type: 'BASIC',
        adjList: []
      },
      nodeBranch2: {
        label: 'Branch 2',
        value: 55,
        type: 'BASIC',
        adjList: ['nodeBranch3']
      },
      nodeBranch3: {
        label: 'Branch 3',
        value: 25,
        type: 'BASIC',
        adjList: []
      }
    }

    DropOutGraphComp = <DropOutGraph jsonData={NodeMap} />

    wrapper = shallow(DropOutGraphComp)
  })

  it('should renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should contain correct classes', () => {
    expect(wrapper.find('.chartcontainer')).toHaveLength(1)
  })

  it('should contain correct components', () => {
    expect(wrapper.find('VerticalBarChart')).toHaveLength(5)
    expect(wrapper.find('DiffIndicator')).toHaveLength(4)
  })
})
