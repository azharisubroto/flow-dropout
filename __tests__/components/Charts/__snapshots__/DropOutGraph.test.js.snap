// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropOutGraph /> should renders correctly 1`] = `ShallowWrapper {}`;
